File: 2967-minimum-cost-to-make-array-equalindromic/2967-minimum-cost-to-make-array-equalindromic.py
class Solution:
    def minimumCost(self, nums: List[int]) -> int:
        
        nums = sorted(nums)
        
        def isPalindrome(num):
            str_num = str(num)
            i, j = 0, len(str_num)-1
            while i < j:
                if str_num[i] != str_num[j]:
                    return False
                i, j = i+1, j-1
            return True
        
        median = nums[len(nums)//2]
        lower_pal, higher_pal = median, median
        while lower_pal > 0:
            if isPalindrome(lower_pal):
                break
            lower_pal-=1
            
        while higher_pal < 10e19:
            if isPalindrome(higher_pal):
                break
            higher_pal+=1
        
        cost1, cost2 = 0, 0
        for i in range(len(nums)):
            cost1 += abs(nums[i] - lower_pal)
            cost2 += abs(nums[i] - higher_pal)
            
        return min(cost1, cost2)

File: 2967-minimum-cost-to-make-array-equalindromic/2967-minimum-cost-to-make-array-equalindromic.java
class Solution {
    public long minimumCost(int[] nums) {
        Arrays.sort(nums);
        
        int median = nums[nums.length/2];
        
        int median_left = median;
        int median_right = median;

        // find closest palidrome to left and right of medium
        
        while(!palindrome(median_left)){
            median_left--;
        }
        while(!palindrome(median_right)){
            median_right++;
        }

        // add to cost and return min_cost
        
        long cost1 = 0, cost2 = 0;
        for(int i=0;i<nums.length;i++){
            cost1 += Math.abs(nums[i] - median_left);
            cost2 += Math.abs(nums[i] - median_right);
        }
        return Math.min(cost1, cost2);
        
    }
    boolean palindrome(long n){
        long orig = n;
        long num = 0;
        while(orig > 0){
            num = num*10 + orig%10;
            orig /= 10;
        }
        return num == n;
    }
}